File: tests/unit/test_wasm_coroutine.zig
//! Zig binding tests for the stackless WASM coroutine/future core.
//!
//! `zig test` compiles src/libsys/wasm/coroutine_wasm.c for the build host and
//! links it against src/libc/zig/coroutine.zig, so the core exercised here is
//! the same C implementation every other language suite runs, executed as host
//! code rather than as wasm. What is unique is the BINDING: coroutine.zig
//! re-declares the C types as `extern struct`s and enums, and a wrong field
//! order, width or tag type surfaces only as a garbled value.
//!
//! The scenarios mirror tests/unit/test_wasm_coroutine.c.
const std = @import("std");
const coroutine = @import("coroutine");

const Waiter = struct {
    pc: i32 = 0,
    future: *coroutine.Future,
    value: usize = 0,
    status: i32 = -1,
};

fn waiterResume(user: ?*anyopaque, out: *usize) callconv(.c) i32 {
    const state: *Waiter = @ptrCast(@alignCast(user.?));
    if (state.pc == 0) {
        state.pc = 1;
        switch (state.future.awaitValue()) {
            .pending => return coroutine.TaskResult.yielded,
            .ready => |value| state.value = value,
            .failed => |status| return status,
            .invalid => return -1,
        }
    }
    switch (state.future.awaitValue()) {
        .pending => return coroutine.TaskResult.yielded,
        .ready => |value| {
            state.value = value;
            state.status = 0;
            out.* = value;
            return coroutine.TaskResult.complete;
        },
        .failed => |status| return status,
        .invalid => return -1,
    }
}

fn increment(_: ?*anyopaque, value: usize, out: *usize) callconv(.c) i32 {
    out.* = value + 1;
    return 0;
}

test "WASM Zig coroutine methods use the shared future core" {
    var runtime: coroutine.Runtime = .{};
    runtime.init();
    var future: coroutine.Future = .{};
    var promise: coroutine.Promise = .{};
    future.init(&promise);
    var task: coroutine.Coroutine = .{};
    var waiter = Waiter{ .future = &future };
    try std.testing.expect(task.start(&runtime, waiterResume, &waiter) != null);
    try std.testing.expectEqual(@as(i32, 1), runtime.run());
    try std.testing.expect(promise.resolve(42));
    try std.testing.expectEqual(@as(i32, 1), runtime.run());
    try std.testing.expectEqual(@as(usize, 42), waiter.value);
    var joined: i32 = 0;
    try std.testing.expectEqual(@as(i32, 0), task.join(&joined));
    try std.testing.expectEqual(@as(i32, 42), joined);

    var source: coroutine.Future = .{};
    var source_promise: coroutine.Promise = .{};
    source.init(&source_promise);
    var continuation: coroutine.Continuation = .{};
    const child = source.then(&runtime, &continuation, increment, null, null) orelse unreachable;
    try std.testing.expect(source_promise.resolve(20));
    try std.testing.expect(child.poll() == null);
    try std.testing.expectEqual(@as(i32, 0), runtime.run());
    switch (child.poll().?) {
        .ready => |value| try std.testing.expectEqual(@as(usize, 21), value),
        .failed => unreachable,
    }

    var first: coroutine.Future = .{};
    var first_promise: coroutine.Promise = .{};
    first.init(&first_promise);
    var second: coroutine.Future = .{};
    var second_promise: coroutine.Promise = .{};
    second.init(&second_promise);
    var group: coroutine.FutureGroup = .{};
    var continuations = [_]coroutine.Continuation{ .{}, .{} };
    const inputs = [_]*coroutine.Future{ &first, &second };
    try std.testing.expect(group.race(&runtime, &inputs, &continuations) != null);
    try std.testing.expect(second_promise.resolve(7));
    try std.testing.expect(first_promise.reject(-2));
    try std.testing.expectEqual(@as(i32, 0), runtime.run());
}

// ---------------------------------------------------------------------------
// FFI smoke, then the full battery. Mirrors tests/unit/test_wasm_coroutine.c;
// this suite links the same coroutine_wasm.c, so what it uniquely checks is the
// BINDING: that the extern struct field order and types agree with C.
// ---------------------------------------------------------------------------

test "FFI layout round trips through the binding" {
    var runtime: coroutine.Runtime = .{};
    runtime.init();
    var future: coroutine.Future = .{};
    var promise: coroutine.Promise = .{};
    future.init(&promise);

    try std.testing.expect(future.poll() == null);
    try std.testing.expect(promise.resolve(0xABCD1234));
    // A value stored by C is read back through the Zig field declarations; a
    // wrong field order or type would garble it.
    switch (future.poll().?) {
        .ready => |value| try std.testing.expectEqual(@as(usize, 0xABCD1234), value),
        .failed => unreachable,
    }
    try std.testing.expect(!promise.resolve(1));

    var rejected: coroutine.Future = .{};
    var rejected_promise: coroutine.Promise = .{};
    rejected.init(&rejected_promise);
    try std.testing.expect(!rejected_promise.reject(0));
    try std.testing.expect(rejected_promise.reject(-77));
    switch (rejected.poll().?) {
        .ready => unreachable,
        .failed => |status| try std.testing.expectEqual(@as(i32, -77), status),
    }
    try std.testing.expectEqual(@as(i32, 0), runtime.run());
}

const CallbackState = struct { calls: u32 = 0 };

fn countingIncrement(user: ?*anyopaque, value: usize, out: *usize) callconv(.c) i32 {
    const state: *CallbackState = @ptrCast(@alignCast(user.?));
    state.calls += 1;
    out.* = value + 1;
    return 0;
}

fn recoverCallback(user: ?*anyopaque, status: i32, out: *usize) callconv(.c) i32 {
    const state: *CallbackState = @ptrCast(@alignCast(user.?));
    state.calls += 1;
    if (status >= 0) return -1;
    out.* = 55;
    return 0;
}

test "callbacks are deferred, even on an already-settled future" {
    var runtime: coroutine.Runtime = .{};
    runtime.init();
    var state: CallbackState = .{};

    var source: coroutine.Future = .{};
    var source_promise: coroutine.Promise = .{};
    source.init(&source_promise);
    var plus_one: coroutine.Continuation = .{};
    const child = source.then(&runtime, &plus_one, countingIncrement, null, &state).?;
    try std.testing.expect(source_promise.resolve(20));
    // Resolving does not run the callback; the runtime does.
    try std.testing.expectEqual(@as(u32, 0), state.calls);
    try std.testing.expectEqual(@as(i32, 0), runtime.run());
    try std.testing.expectEqual(@as(u32, 1), state.calls);
    switch (child.poll().?) {
        .ready => |value| try std.testing.expectEqual(@as(usize, 21), value),
        .failed => unreachable,
    }

    // The error handler recovers, so the child SUCCEEDS.
    var rejected: coroutine.Future = .{};
    var rejected_promise: coroutine.Promise = .{};
    rejected.init(&rejected_promise);
    var recover_continuation: coroutine.Continuation = .{};
    const recovered = rejected.then(&runtime, &recover_continuation, null, recoverCallback, &state).?;
    try std.testing.expect(rejected_promise.reject(-23));
    try std.testing.expect(!rejected_promise.reject(-24));
    try std.testing.expectEqual(@as(i32, 0), runtime.run());
    switch (recovered.poll().?) {
        .ready => |value| try std.testing.expectEqual(@as(usize, 55), value),
        .failed => unreachable,
    }

    // Registering on an ALREADY-SETTLED future must still defer. Every other
    // case registers before settling, so this is the only one covering that
    // branch.
    var settled: coroutine.Future = .{};
    var settled_promise: coroutine.Promise = .{};
    settled.init(&settled_promise);
    try std.testing.expect(settled_promise.resolve(70));
    state.calls = 0;
    var late: coroutine.Continuation = .{};
    const late_child = settled.then(&runtime, &late, countingIncrement, null, &state).?;
    try std.testing.expectEqual(@as(u32, 0), state.calls);
    try std.testing.expectEqual(@as(i32, 0), runtime.run());
    try std.testing.expectEqual(@as(u32, 1), state.calls);
    switch (late_child.poll().?) {
        .ready => |value| try std.testing.expectEqual(@as(usize, 71), value),
        .failed => unreachable,
    }
}

const ManyWaiter = struct {
    future: *coroutine.Future,
    status: i32 = -1,
    value: usize = 0,
};

fn manyWaiterResume(user: ?*anyopaque, out: *usize) callconv(.c) i32 {
    const state: *ManyWaiter = @ptrCast(@alignCast(user.?));
    switch (state.future.awaitValue()) {
        .pending => return coroutine.TaskResult.yielded,
        .ready => |value| {
            state.status = 0;
            state.value = value;
            out.* = value;
            return coroutine.TaskResult.complete;
        },
        .failed => |status| {
            state.status = status;
            return status;
        },
        .invalid => return -1,
    }
}

test "one settle wakes every waiter" {
    const waiters = 4;
    var failing: u32 = 0;
    while (failing < 2) : (failing += 1) {
        var runtime: coroutine.Runtime = .{};
        runtime.init();
        var future: coroutine.Future = .{};
        var promise: coroutine.Promise = .{};
        future.init(&promise);

        var states: [waiters]ManyWaiter = undefined;
        var coroutines: [waiters]coroutine.Coroutine = undefined;
        for (0..waiters) |i| {
            states[i] = .{ .future = &future };
            coroutines[i] = .{};
            try std.testing.expect(coroutines[i].start(&runtime, manyWaiterResume, &states[i]) != null);
        }
        // The wait list is spliced whole at settle time; with a single waiter
        // that loop runs once with next == null, so several waiters on one
        // future are what exercises it.
        try std.testing.expectEqual(@as(i32, waiters), runtime.run());
        if (failing == 0) {
            try std.testing.expect(promise.resolve(77));
        } else {
            try std.testing.expect(promise.reject(-31));
        }
        try std.testing.expectEqual(@as(i32, waiters), runtime.run());
        for (0..waiters) |i| {
            if (failing == 0) {
                try std.testing.expectEqual(@as(i32, 0), states[i].status);
                try std.testing.expectEqual(@as(usize, 77), states[i].value);
            } else {
                try std.testing.expectEqual(@as(i32, -31), states[i].status);
            }
        }
    }
}

const StressState = struct { remaining: u32, completed: *u32 };

fn stressResume(user: ?*anyopaque, out: *usize) callconv(.c) i32 {
    const state: *StressState = @ptrCast(@alignCast(user.?));
    if (state.remaining > 0) {
        state.remaining -= 1;
        return coroutine.TaskResult.yielded;
    }
    state.completed.* += 1;
    out.* = 0;
    return coroutine.TaskResult.complete;
}

test "round robin is fair at scale" {
    const count = 8;
    const yields = 4;
    var runtime: coroutine.Runtime = .{};
    runtime.init();
    var completed: u32 = 0;
    var states: [count]StressState = undefined;
    var coroutines: [count]coroutine.Coroutine = undefined;
    for (0..count) |i| {
        states[i] = .{ .remaining = yields, .completed = &completed };
        coroutines[i] = .{};
        try std.testing.expect(coroutines[i].start(&runtime, stressResume, &states[i]) != null);
    }
    // The exact resume count is the schedule.
    try std.testing.expectEqual(@as(i32, count * (yields + 1)), runtime.run());
    try std.testing.expectEqual(@as(u32, count), completed);
}

test "race: first to settle wins from any position, losers discarded" {
    for (0..3) |winner| {
        var runtime: coroutine.Runtime = .{};
        runtime.init();
        var futures: [3]coroutine.Future = .{ .{}, .{}, .{} };
        var promises: [3]coroutine.Promise = .{ .{}, .{}, .{} };
        for (0..3) |i| futures[i].init(&promises[i]);
        var continuations: [3]coroutine.Continuation = .{ .{}, .{}, .{} };
        const inputs = [_]*coroutine.Future{ &futures[0], &futures[1], &futures[2] };
        var group: coroutine.FutureGroup = .{};
        const result = group.race(&runtime, &inputs, &continuations).?;

        const expected: usize = 100 + winner;
        try std.testing.expect(promises[winner].resolve(expected));
        // The rest settle before the drain, so they are losers that were
        // already enqueued rather than merely pending.
        var i: usize = 3;
        while (i > 0) {
            i -= 1;
            if (i == winner) continue;
            try std.testing.expect(promises[i].resolve(900 + i));
        }
        try std.testing.expectEqual(@as(i32, 0), runtime.run());
        switch (result.poll().?) {
            .ready => |value| try std.testing.expectEqual(expected, value),
            .failed => unreachable,
        }
        try std.testing.expect(group.settled);
        try std.testing.expect(!group.active);
        // The losers were DISCARDED, not merely outvoted: exactly one group
        // callback ran. A runtime that leaves them queued and lets them run --
        // their resolve refused because the group already settled -- is
        // otherwise indistinguishable.
        try std.testing.expectEqual(@as(usize, 1), group.completed);
        for (0..3) |j| try std.testing.expect(futures[j].continuations == null);
        try std.testing.expectEqual(@as(i32, 0), runtime.run());
    }
}

test "all fails fast from any position, survivors released" {
    for (0..3) |rejecter| {
        var runtime: coroutine.Runtime = .{};
        runtime.init();
        var futures: [3]coroutine.Future = .{ .{}, .{}, .{} };
        var promises: [3]coroutine.Promise = .{ .{}, .{}, .{} };
        for (0..3) |i| futures[i].init(&promises[i]);
        var continuations: [3]coroutine.Continuation = .{ .{}, .{}, .{} };
        var values: [3]usize = .{ 0, 0, 0 };
        const inputs = [_]*coroutine.Future{ &futures[0], &futures[1], &futures[2] };
        var group: coroutine.FutureGroup = .{};
        const result = group.all(&runtime, &inputs, &values, &continuations).?;

        const expected: i32 = -60 - @as(i32, @intCast(rejecter));
        try std.testing.expect(promises[rejecter].reject(expected));
        try std.testing.expectEqual(@as(i32, 0), runtime.run());
        switch (result.poll().?) {
            .ready => unreachable,
            .failed => |status| try std.testing.expectEqual(expected, status),
        }
        try std.testing.expect(group.settled);
        try std.testing.expect(!group.active);
        try std.testing.expectEqual(@as(usize, 1), group.completed);
        for (0..3) |j| {
            if (j != rejecter) try std.testing.expect(futures[j].continuations == null);
        }
        // A survivor settling afterwards is inert.
        for (0..3) |j| {
            if (j == rejecter) continue;
            try std.testing.expect(promises[j].resolve(999));
        }
        try std.testing.expectEqual(@as(i32, 0), runtime.run());
        switch (result.poll().?) {
            .ready => unreachable,
            .failed => |status| try std.testing.expectEqual(expected, status),
        }
    }
}

File: tests/unit/test_wasm_coroutine.go
// test_wasm_coroutine.go - the Go binding for the shared C coroutine runtime.
//
// This is the one binding that could not be tested the way the others are. The
// Rust and Zig suites run on the HOST, linking coroutine_wasm.c directly, which
// works because their #[repr(C)] / extern struct layouts are computed per
// target. Go cannot express the C layout at all: it models every record as an
// opaque [N]uint32 blob sized for wasm32, so a host test would exercise 64-bit
// pointers and a different ABI entirely. It has to be built for wasm32 and run
// there, which is what tests/unit/go/run_go_test.mjs does.
//
// Go also carries an FFI layer the others do not, and it is the part most
// likely to break:
//
//   - Callback is a raw uintptr, because "Go has no portable representation for
//     a C function pointer", so coroutines route through the C trampoline in
//     go_coroutine_trampoline.c with a caller-assigned task id.
//   - Go closures reach the runtime through a fixed 32-slot registry keyed by
//     callback id.
//
// Neither had any coverage. The scenarios below mirror
// tests/unit/test_wasm_coroutine.c and add the two FFI-specific ones.
//
// Each case returns 0 or a marker; runTests returns the first non-zero.

package main

import "unsafe"

// wasmos.go owns main() and dispatches to Main(); the test entry point is the
// exported runTests below.
func Main(args []string) int32 { return 0 }

// ---------------------------------------------------------------- FFI smoke

// The blob sizes are pinned at compile time by tests/unit/test_go_abi_sizes.c.
// This checks the other half: that values written by C through those pointers
// are read back correctly from Go, i.e. the field OFFSETS agree too. A size can
// be right while the layout is skewed.
func testFFISmoke() int32 {
	var rt Runtime
	rt.Init()

	var future Future
	var promise Promise
	future.Init(&promise)

	// Pending: C reports not-settled and leaves the outputs alone.
	if _, _, settled := future.Poll(); settled {
		return 110
	}
	// A value written by C is read back intact through the blob.
	if !promise.Resolve(0xABCD1234) {
		return 111
	}
	value, status, settled := future.Poll()
	if !settled || status != 0 || value != 0xABCD1234 {
		return 112
	}
	// A settled future refuses to settle again, so the state field is real.
	if promise.Resolve(1) {
		return 113
	}
	// Rejection carries a negative status through the same path.
	var rejected Future
	var rejectedPromise Promise
	rejected.Init(&rejectedPromise)
	if rejectedPromise.Reject(0) {
		return 114
	}
	if !rejectedPromise.Reject(-77) {
		return 115
	}
	if _, status, settled = rejected.Poll(); !settled || status != -77 {
		return 116
	}
	// An empty runtime drains to zero rather than trapping.
	if rt.Run() != 0 {
		return 117
	}
	return 0
}

// ------------------------------------------------------- trampoline tasks

// Coroutines reach Go through wasmos_go_coroutine_resume, dispatched on a
// caller-assigned task id. The registry below is the test's own; the binding
// only carries the id.
type goTask struct {
	pc     int32
	events *[]uint32
	future *Future
	status int32
	value  uintptr
	result uintptr
	mode   int32
}

const (
	modeYield = iota
	modeWait
	modeResolve
	modeStress
)

var goTasks [64]*goTask
var goResolvePromise *Promise
var goStressCompleted int32

//go:export wasmos_go_coroutine_resume
func wasmosGoCoroutineResume(taskID uint32, out *uintptr) int32 {
	if taskID == 0 || int(taskID) > len(goTasks) || goTasks[taskID-1] == nil {
		return -1
	}
	t := goTasks[taskID-1]
	switch t.mode {
	case modeYield:
		if t.pc == 0 {
			*t.events = append(*t.events, 1)
			t.pc = 1
			return TaskYielded
		}
		*t.events = append(*t.events, 3)
		*out = 7
		return TaskComplete
	case modeWait:
		t.value, t.status = t.future.Await()
		if t.status == AwaitPending {
			return TaskYielded
		}
		*out = t.value
		return t.status
	case modeResolve:
		if !goResolvePromise.Resolve(42) {
			return -1
		}
		*out = 42
		return TaskComplete
	case modeStress:
		if t.pc > 0 {
			t.pc--
			return TaskYielded
		}
		goStressCompleted++
		*out = 0
		return TaskComplete
	}
	return -1
}

func registerTask(id uint32, t *goTask) { goTasks[id-1] = t }

// ------------------------------------------------------------------ battery

func testYieldAwaitAndJoin() int32 {
	for i := range goTasks {
		goTasks[i] = nil
	}
	var rt Runtime
	rt.Init()
	var source Future
	var promise Promise
	source.Init(&promise)
	goResolvePromise = &promise

	events := make([]uint32, 0, 4)
	yielder := &goTask{mode: modeYield, events: &events}
	waiter := &goTask{mode: modeWait, future: &source}
	resolver := &goTask{mode: modeResolve}
	registerTask(1, yielder)
	registerTask(2, waiter)
	registerTask(3, resolver)

	var first, waiterCo, resolverCo Coroutine
	if first.StartTask(&rt, 1) == nil {
		return 210
	}
	if waiterCo.StartTask(&rt, 2) == nil {
		return 211
	}
	if resolverCo.StartTask(&rt, 3) == nil {
		return 212
	}
	// Five resumes: yielder twice, waiter twice, resolver once.
	if rt.Run() != 5 {
		return 213
	}
	if len(events) != 2 || events[0] != 1 || events[1] != 3 {
		return 214
	}
	if waiter.status != 0 || waiter.value != 42 {
		return 215
	}
	result, status := first.Join()
	if status != 0 || result != 7 {
		return 216
	}
	if promise.Resolve(9) {
		return 217
	}
	return 0
}

func testMultipleWaitersAndJoiners() int32 {
	for i := range goTasks {
		goTasks[i] = nil
	}
	var rt Runtime
	rt.Init()
	var source Future
	var promise Promise
	source.Init(&promise)

	const waiters = 4
	tasks := make([]*goTask, waiters)
	cos := make([]Coroutine, waiters)
	for i := 0; i < waiters; i++ {
		tasks[i] = &goTask{mode: modeWait, future: &source}
		registerTask(uint32(i+1), tasks[i])
		if cos[i].StartTask(&rt, uint32(i+1)) == nil {
			return 310
		}
	}
	// All four park on the first drain.
	if rt.Run() != waiters {
		return 311
	}
	if !promise.Resolve(77) {
		return 312
	}
	// Settling wakes every one of them, not just the head of the wait list.
	if rt.Run() != waiters {
		return 313
	}
	for i := 0; i < waiters; i++ {
		if tasks[i].status != 0 || tasks[i].value != 77 {
			return 314
		}
	}
	return 0
}

func testSchedulerStress() int32 {
	for i := range goTasks {
		goTasks[i] = nil
	}
	var rt Runtime
	rt.Init()
	const count = 8
	const yields = 4
	goStressCompleted = 0
	cos := make([]Coroutine, count)
	for i := 0; i < count; i++ {
		registerTask(uint32(i+1), &goTask{mode: modeStress, pc: yields})
		if cos[i].StartTask(&rt, uint32(i+1)) == nil {
			return 410
		}
	}
	// The exact resume count is the schedule.
	if rt.Run() != count*(yields+1) {
		return 411
	}
	if goStressCompleted != count {
		return 412
	}
	return 0
}

func testChainsAndDeferral() int32 {
	var rt Runtime
	rt.Init()
	calls := 0

	var source Future
	var promise Promise
	source.Init(&promise)
	var plusOne Continuation
	child := source.ThenGo(&rt, &plusOne, func(v uintptr) (uintptr, int32) {
		calls++
		return v + 1, 0
	}, nil)
	if child == nil {
		return 510
	}
	if !promise.Resolve(20) {
		return 511
	}
	// Deferred: resolving does not run the callback, the runtime does.
	if calls != 0 {
		return 512
	}
	if rt.Run() != 0 || calls != 1 {
		return 513
	}
	if v, s, ok := child.Poll(); !ok || s != 0 || v != 21 {
		return 514
	}

	// Error recovery through the Go error dispatcher.
	var rejected Future
	var rejectedPromise Promise
	rejected.Init(&rejectedPromise)
	var recover Continuation
	child = rejected.CatchGo(&rt, &recover, func(status int32) (uintptr, int32) {
		if status >= 0 {
			return 0, -1
		}
		return 55, 0
	})
	if child == nil {
		return 520
	}
	if !rejectedPromise.Reject(-23) || rejectedPromise.Reject(-24) {
		return 521
	}
	if rt.Run() != 0 {
		return 522
	}
	if v, s, ok := child.Poll(); !ok || s != 0 || v != 55 {
		return 523
	}

	// Registering on an ALREADY-SETTLED future still defers.
	var settled Future
	var settledPromise Promise
	settled.Init(&settledPromise)
	if !settledPromise.Resolve(70) {
		return 530
	}
	late := 0
	var lateContinuation Continuation
	child = settled.ThenGo(&rt, &lateContinuation, func(v uintptr) (uintptr, int32) {
		late++
		return v + 1, 0
	}, nil)
	if child == nil || late != 0 {
		return 531
	}
	if rt.Run() != 0 || late != 1 {
		return 532
	}
	if v, _, ok := child.Poll(); !ok || v != 71 {
		return 533
	}
	return 0
}

// Race with three candidates that all settle, once per winning position, with
// the losers required to be discarded rather than merely outvoted.
func raceWinnerCase(winner int, base int32) int32 {
	var rt Runtime
	rt.Init()
	futures := make([]Future, 3)
	promises := make([]Promise, 3)
	inputs := make([]*Future, 3)
	continuations := make([]Continuation, 3)
	for i := 0; i < 3; i++ {
		futures[i].Init(&promises[i])
		inputs[i] = &futures[i]
	}
	var group FutureGroup
	result := group.Race(&rt, inputs, continuations)
	if result == nil {
		return base + 0
	}
	expected := uintptr(100 + winner)
	if !promises[winner].Resolve(expected) {
		return base + 1
	}
	for i := 2; i >= 0; i-- {
		if i == winner {
			continue
		}
		if !promises[i].Resolve(uintptr(900 + i)) {
			return base + 2
		}
	}
	if rt.Run() != 0 {
		return base + 3
	}
	v, s, ok := result.Poll()
	if !ok || s != 0 || v != expected {
		return base + 4
	}
	// Losers settling later must not disturb the result.
	if rt.Run() != 0 {
		return base + 5
	}
	if v, _, ok = result.Poll(); !ok || v != expected {
		return base + 6
	}
	return 0
}

func testRaceEveryWinner() int32 {
	if rc := raceWinnerCase(0, 600); rc != 0 {
		return rc
	}
	if rc := raceWinnerCase(1, 700); rc != 0 {
		return rc
	}
	return raceWinnerCase(2, 800)
}

func testAll() int32 {
	var rt Runtime
	rt.Init()
	futures := make([]Future, 3)
	promises := make([]Promise, 3)
	inputs := make([]*Future, 3)
	continuations := make([]Continuation, 3)
	values := make([]uintptr, 3)
	for i := 0; i < 3; i++ {
		futures[i].Init(&promises[i])
		inputs[i] = &futures[i]
	}
	var group FutureGroup
	result := group.All(&rt, inputs, values, continuations)
	if result == nil {
		return 910
	}
	// Settled out of order; values must land in INPUT order.
	if !promises[2].Resolve(3) || !promises[0].Resolve(1) || !promises[1].Resolve(2) {
		return 911
	}
	if rt.Run() != 0 {
		return 912
	}
	v, s, ok := result.Poll()
	if !ok || s != 0 || v != uintptr(unsafe.Pointer(&values[0])) {
		return 913
	}
	if values[0] != 1 || values[1] != 2 || values[2] != 3 {
		return 914
	}
	return 0
}

// The 32-slot Go callback registry. Every ThenGo takes a slot; before slots
// were reclaimed on dispatch the table filled permanently and ThenGo returned
// nil forever after, so a long-running service stopped chaining after its 32nd
// callback with no diagnostic.
func testCallbackRegistryReclaim() int32 {
	var rt Runtime
	rt.Init()
	const rounds = 32*3 + 5
	for round := 0; round < rounds; round++ {
		var source Future
		var promise Promise
		source.Init(&promise)
		var continuation Continuation
		fired := false
		child := source.ThenGo(&rt, &continuation, func(v uintptr) (uintptr, int32) {
			fired = true
			return v + 1, 0
		}, nil)
		if child == nil {
			// Exactly the failure the reclaim exists to prevent.
			return 1000 + int32(round)
		}
		if !promise.Resolve(uintptr(round)) {
			return 1100 + int32(round)
		}
		if rt.Run() != 0 || !fired {
			return 1200 + int32(round)
		}
		if v, s, ok := child.Poll(); !ok || s != 0 || v != uintptr(round+1) {
			return 1300 + int32(round)
		}
	}
	// A rejected chain releases through the error path, not only the success one.
	for round := 0; round < rounds; round++ {
		var source Future
		var promise Promise
		source.Init(&promise)
		var continuation Continuation
		child := source.CatchGo(&rt, &continuation, func(status int32) (uintptr, int32) {
			return 5, 0
		})
		if child == nil {
			return 1400 + int32(round)
		}
		if !promise.Reject(-3) {
			return 1500 + int32(round)
		}
		if rt.Run() != 0 {
			return 1600 + int32(round)
		}
		if v, s, ok := child.Poll(); !ok || s != 0 || v != 5 {
			return 1700 + int32(round)
		}
	}
	return 0
}

//go:wasmexport runTests
func runTests() int32 {
	if rc := testFFISmoke(); rc != 0 {
		return rc
	}
	if rc := testYieldAwaitAndJoin(); rc != 0 {
		return rc
	}
	if rc := testMultipleWaitersAndJoiners(); rc != 0 {
		return rc
	}
	if rc := testSchedulerStress(); rc != 0 {
		return rc
	}
	if rc := testChainsAndDeferral(); rc != 0 {
		return rc
	}
	if rc := testRaceEveryWinner(); rc != 0 {
		return rc
	}
	if rc := testAll(); rc != 0 {
		return rc
	}
	return testCallbackRegistryReclaim()
}
